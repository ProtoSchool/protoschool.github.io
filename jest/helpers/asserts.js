// ** SNAPSHOT FILTERING ***
// prevent snapshots for logs that change on each instance, such as shuffled quiz options

const excludedPhrases = [
  "Cool! Here's what we get when we mix up the order of the answer choices:"
]

// returns true if log includes a string from the excludedPhrases array
function isExcluded (log) {
  return excludedPhrases.some(phrase => log.includes(phrase))
}

// add log to snapshot UNLESS it contains an excluded phrase
export function assertLogSnapshot (log) {
  if (!isExcluded(log)) {
    expect(log).toMatchSnapshot() // expect if log isn't excluded from assertion
  }
  // do nothing if log is excluded from assertion
}
