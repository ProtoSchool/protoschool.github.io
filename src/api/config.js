/* eslint-disable no-unused-expressions */
export const staticPath = process.env.DATA_STATIC_PATH || 'src/static'
export const tutorialsPath = process.env.DATA_TUTORIALS_PATH || 'src/tutorials'

const boilerplatesPath = process.env.LESSON_BOILERPLATES_PATH || `${tutorialsPath}/boilerplates`

export const boilerplates = {
  path: boilerplatesPath,
  markdownPath: `${boilerplatesPath}/boilerplate.md`,
  challengeMarkdownPath: `${boilerplatesPath}/boilerplate-challenge.md`,
  multipleChoiceJsPath: `${boilerplatesPath}/boilerplate-multiple-choice.js`,
  codeJsPath: `${boilerplatesPath}/boilerplate-code.js`,
  fileUploadJsPath: `${boilerplatesPath}/boilerplate-file-upload.js`
}
