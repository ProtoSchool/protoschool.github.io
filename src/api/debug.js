export default (process.env.NODE_ENV === 'development' || process.env.NODE_ENV === 'test') && process.env.DEBUG
