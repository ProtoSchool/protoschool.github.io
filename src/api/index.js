/**
 * API for ProtoSchool.
 *
 * TODO: Make API universal https://github.com/ProtoSchool/protoschool.github.io/issues/589
 *  - make all scripts use this API
 *  - change it to be universal so the website can also use it
 *   - ATM we have deps on fs.readFile etc
 *
 * @module api
 */
module.exports = {
  projects: require('./modules/projects'),
  lessons: require('./modules/lessons'),
  courses: require('./modules/courses').default,
  resources: require('./modules/resources'),
  tutorials: require('./modules/tutorials'),
  _config: require('./config')
}
