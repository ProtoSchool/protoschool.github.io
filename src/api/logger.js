export const debug = (...args) => {
  console.log('⚙️ api', ...args)
}

export const createLogGroup = module => functionMethod => `[${module}.${functionMethod}()]`
