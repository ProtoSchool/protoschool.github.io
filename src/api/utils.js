import { writeFileSync } from 'fs'
import { resolve } from 'path'
import { staticPath } from './config.js'

const correctedCases = {
  api: 'API',
  cid: 'CID',
  of: 'of',
  a: 'a'
}

export function writeStaticFile(staticFileName, data) {
  writeFileSync(resolve(staticPath, staticFileName), JSON.stringify(data, null, 2))
}

export function deriveShortname(path) {
  return path.split('-').map(word => (
    correctedCases[word] ? correctedCases[word] : (word.charAt(0).toUpperCase() + word.slice(1))
  )).join(' ')
}
