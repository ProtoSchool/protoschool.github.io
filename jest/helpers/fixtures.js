import { lessons as _lessons, resources as _resources, courses, tutorials } from '../../src/api'

// Creators: create data (tutorials, lessons)
export function createTutorial (config = { override: {}, lessons: 0, resources: 0 }) {
  const { tutorial, lessons, resources } = generateTutorial(config)
  const createdTutorial = tutorials.create(tutorial)

  for (let i = 0; i < lessons.length; ++i) {
    _lessons.create(tutorials.get(createdTutorial.id), lessons[i])
  }

  resources.forEach(resource => {
    _resources.add(createdTutorial.id, resource)
  })

  return tutorials.get(createdTutorial.id)
}

// Generators: generate data to be used

export function generateTutorial (config = { override: {}, lessons: 0, resources: 0, lessonOverride: {} }) {
  const suffix = tutorials.getNextTutorialId()

  const tutorial = {
    title: `New Tutorial (${suffix})`,
    url: `new-tut-${suffix}`,
    project: 'libp2p',
    description: 'New tutorial description',
    ...config.override
  }

  const lessons = new Array(config.lessons).fill().map((_, i) => (
    generateLesson({
      override: {
        title: `Lesson ${i + 1}`,
        ...config.lessonOverride
      }
    }).lesson
  ))

  const resources = new Array(config.resources).fill().map((_, i) => (
    generateResource({
      override: {
        title: `Resource ${i + 1}`,
        link: `https://resource${i + 1}.com`
      }
    }).resource
  ))

  return {
    tutorial,
    lessons,
    resources,
    expected: { ...tutorial, lessons, resources }
  }
}

export function generateLesson ({ createTutorial = false, override = {} } = {}) {
  const lesson = {
    title: 'Lesson',
    type: 'text',
    ...override
  }
  let tutorial

  if (createTutorial) {
    tutorial = tutorials.create(generateTutorial().tutorial)
    courses.add(tutorial.id)
  }

  return {
    lesson,
    tutorial,
    expected: { lesson, tutorial }
  }
}

export function generateResource ({ createTutorial, override = {} } = {}) {
  const resource = {
    title: 'Resource',
    link: 'https://resource.com',
    description: 'Resoure description',
    type: 'demo',
    ...override
  }
  let tutorial

  if (createTutorial) {
    tutorial = tutorials.create(generateTutorial().tutorial)
    courses.add(tutorial.id)
  }

  return {
    resource,
    tutorial,
    expected: { resource, tutorial }
  }
}
