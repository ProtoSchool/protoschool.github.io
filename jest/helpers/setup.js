import { courses, tutorials } from '../../src/api'

export function restoreData (lastTutorialId) {
  const newLastTutorialId = tutorials.list.getLatest().id

  // delete all new tutorials
  for (let id = lastTutorialId + 1; id <= newLastTutorialId; id++) {
    tutorials.remove(id)
    courses.remove(id)
  }
}
