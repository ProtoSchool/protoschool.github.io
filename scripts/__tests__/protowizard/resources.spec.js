import { tutorials } from '../../../src/api'
import { restoreData } from '../../../jest/helpers/setup'
import { generateTutorial, generateResource } from '../../../jest/helpers/fixtures'
import { assertNewTutorial, assertNewResource } from '../helpers/asserts'
import { protowizard } from '../helpers/runners'

describe('protowizard', () => {
  let lastTutorialId

  beforeAll(() => {
    lastTutorialId = tutorials.list.getLatest().id
  })

  afterEach(() => {
    restoreData(lastTutorialId)
  })

  describe('3. create resource', () => {
    test('3.1. should create resource after creating a new tutorial (skips lesson creation)', async () => {
      const { tutorial, resources, expected } = generateTutorial({
        resources: 1
      })

      await protowizard([
        { type: 'resource' },
        { confirm: false }, // no add to latest tutorial
        { tutorialId: 'new' }, // choose to create a new tutorial
        tutorial,
        resources[0],
        { confirm: false }, // no to create another resource
        { confirm: false } // no to create first lesson
      ])

      const result = tutorials.getByUrl(tutorial.url)

      expect(result.resources).toHaveLength(1)

      assertNewTutorial({
        context: { lastTutorialId },
        expected,
        result
      })
    })

    test('3.2. should create a resource and add it to the latest tutorial (skips lesson creation)', async () => {
      const { resource, tutorial, expected } = generateResource({ createTutorial: true })

      await protowizard([
        { type: 'resource' },
        { latestTutorial: true }, // yes add to latest tutorial
        resource,
        { confirm: false }, // no to create another resource
        { confirm: false } // no to create first lesson
      ])

      const result = tutorials.get(tutorial.id)

      expect(result.resources).toHaveLength(1)

      assertNewResource({
        expected,
        result
      })
    })
  })
})
