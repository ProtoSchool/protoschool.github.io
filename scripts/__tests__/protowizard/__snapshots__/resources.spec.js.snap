// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`protowizard 3. create resource 3.1. should create resource after creating a new tutorial (skips lesson creation) 1`] = `"Let's add resources to your tutorial."`;

exports[`protowizard 3. create resource 3.1. should create resource after creating a new tutorial (skips lesson creation) 2`] = `"Let's create the files you need to build your tutorial. I'll ask you a few questions to get started."`;

exports[`protowizard 3. create resource 3.1. should create resource after creating a new tutorial (skips lesson creation) 3`] = `"Thanks! I've created a directory for your tutorial at \`src/tutorials/0011-new-tut-11/\`."`;

exports[`protowizard 3. create resource 3.1. should create resource after creating a new tutorial (skips lesson creation) 4`] = `"To preview your tutorial, first run \`npm start\` in a separate terminal window or tab, then visit this page in your web browser: http://localhost:3000/new-tut-11/"`;

exports[`protowizard 3. create resource 3.1. should create resource after creating a new tutorial (skips lesson creation) 5`] = `"Let's create your first resource!"`;

exports[`protowizard 3. create resource 3.1. should create resource after creating a new tutorial (skips lesson creation) 6`] = `"I've added \\"Resource 1\\" to your resources list."`;

exports[`protowizard 3. create resource 3.1. should create resource after creating a new tutorial (skips lesson creation) 7`] = `
"Okay, sounds like we're done. Here are all the resources now included in \\"New Tutorial (11)\\":
 ‣ Resource 1 (demo)"
`;

exports[`protowizard 3. create resource 3.1. should create resource after creating a new tutorial (skips lesson creation) 8`] = `"To preview your resources page, first run \`npm start\` in a separate terminal window or tab, then visit this page in your web browser: http://localhost:3000/new-tut-11/resources"`;

exports[`protowizard 3. create resource 3.1. should create resource after creating a new tutorial (skips lesson creation) 9`] = `"Need to change something? You can edit your resources in the file \`src/static/tutorials.json\`."`;

exports[`protowizard 3. create resource 3.1. should create resource after creating a new tutorial (skips lesson creation) 10`] = `"Looks like your \\"New Tutorial (11)\\" tutorial doesn't have any lessons yet."`;

exports[`protowizard 3. create resource 3.1. should create resource after creating a new tutorial (skips lesson creation) 11`] = `"Okay, no problem. You can summon me later to add lessons."`;

exports[`protowizard 3. create resource 3.1. should create resource after creating a new tutorial (skips lesson creation) 12`] = `"View the detailed guide to developing tutorials at: https://bit.ly/protoschool-developing"`;

exports[`protowizard 3. create resource 3.2. should create a resource and add it to the latest tutorial (skips lesson creation) 1`] = `"Welcome! I'm the ProtoWizard, and I'm excited to help you build your ProtoSchool tutorial."`;

exports[`protowizard 3. create resource 3.2. should create a resource and add it to the latest tutorial (skips lesson creation) 2`] = `"Let's add resources to your tutorial."`;

exports[`protowizard 3. create resource 3.2. should create a resource and add it to the latest tutorial (skips lesson creation) 3`] = `"Let's create your first resource!"`;

exports[`protowizard 3. create resource 3.2. should create a resource and add it to the latest tutorial (skips lesson creation) 4`] = `"I've added \\"Resource\\" to your resources list."`;

exports[`protowizard 3. create resource 3.2. should create a resource and add it to the latest tutorial (skips lesson creation) 5`] = `
"Okay, sounds like we're done. Here are all the resources now included in \\"New Tutorial (11)\\":
 ‣ Resource (demo)"
`;

exports[`protowizard 3. create resource 3.2. should create a resource and add it to the latest tutorial (skips lesson creation) 6`] = `"To preview your resources page, first run \`npm start\` in a separate terminal window or tab, then visit this page in your web browser: http://localhost:3000/new-tut-11/resources"`;

exports[`protowizard 3. create resource 3.2. should create a resource and add it to the latest tutorial (skips lesson creation) 7`] = `"Need to change something? You can edit your resources in the file \`src/static/tutorials.json\`."`;

exports[`protowizard 3. create resource 3.2. should create a resource and add it to the latest tutorial (skips lesson creation) 8`] = `"Looks like your \\"New Tutorial (11)\\" tutorial doesn't have any lessons yet."`;

exports[`protowizard 3. create resource 3.2. should create a resource and add it to the latest tutorial (skips lesson creation) 9`] = `"Okay, no problem. You can summon me later to add lessons."`;

exports[`protowizard 3. create resource 3.2. should create a resource and add it to the latest tutorial (skips lesson creation) 10`] = `"View the detailed guide to developing tutorials at: https://bit.ly/protoschool-developing"`;
